From: David Daney <david.daney@cavium.com>
Date: Fri, 8 Apr 2016 15:50:26 -0700
Subject: arm64: Move unflatten_device_tree() call earlier.
Git-commit: 3194ac6e66cc7a00c1fa9fecf33a7c376b489497
Patch-mainline: v4.7-rc1
References: fate#319973

In order to extract NUMA information from the device tree, we need to
have the tree in its unflattened form.

Move the call to bootmem_init() in the tail of paging_init() into
setup_arch, and adjust header files so that its declaration is
visible.

Move the unflatten_device_tree() call between the calls to
paging_init() and bootmem_init().  Follow on patches add NUMA handling
to bootmem_init().

Signed-off-by: David Daney <david.daney@cavium.com>
Signed-off-by: Will Deacon <will.deacon@arm.com>
Signed-off-by: Matthias Brugger <mbrugger@suse.com>

---
 arch/arm64/include/asm/mmu.h |    1 +
 arch/arm64/kernel/setup.c    |   12 ++++++++----
 arch/arm64/mm/mm.h           |    1 -
 arch/arm64/mm/mmu.c          |    2 --
 4 files changed, 9 insertions(+), 7 deletions(-)

--- a/arch/arm64/include/asm/mmu.h
+++ b/arch/arm64/include/asm/mmu.h
@@ -29,6 +29,7 @@ typedef struct {
 #define ASID(mm)	((mm)->context.id.counter & 0xffff)
 
 extern void paging_init(void);
+extern void bootmem_init(void);
 extern void __iomem *early_io_map(phys_addr_t phys, unsigned long virt);
 extern void init_mem_pgprot(void);
 extern void create_pgd_mapping(struct mm_struct *mm, phys_addr_t phys,
--- a/arch/arm64/kernel/setup.c
+++ b/arch/arm64/kernel/setup.c
@@ -324,18 +324,22 @@ void __init setup_arch(char **cmdline_p)
 	paging_init();
 	relocate_initrd();
 
+	if (acpi_disabled)
+		unflatten_device_tree();
+
+	bootmem_init();
+
 	kasan_init();
 
 	request_standard_resources();
 
 	early_ioremap_reset();
 
-	if (acpi_disabled) {
-		unflatten_device_tree();
+	if (acpi_disabled)
 		psci_dt_init();
-	} else {
+	else
 		psci_acpi_init();
-	}
+
 	xen_early_init();
 
 	cpu_read_bootcpu_ops();
--- a/arch/arm64/mm/mm.h
+++ b/arch/arm64/mm/mm.h
@@ -1,3 +1,2 @@
-extern void __init bootmem_init(void);
 
 void fixup_init(void);
--- a/arch/arm64/mm/mmu.c
+++ b/arch/arm64/mm/mmu.c
@@ -452,8 +452,6 @@ void __init paging_init(void)
 	/* allocate the zero page. */
 	zero_page = early_alloc(PAGE_SIZE);
 
-	bootmem_init();
-
 	empty_zero_page = virt_to_page(zero_page);
 
 	/* Ensure the zero page is visible to the page table walker */
